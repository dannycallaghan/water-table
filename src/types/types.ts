export type MessageFilters = 'direction' | 'date' | '';

export interface ISortOptions {
  type: string;
  key: string;
  desc: boolean;
}

export interface IPaginationOptions {
  itemsPerPage: number;
  page: number;
}

export interface IMessageDataOptions {
  searchTerm: string;
  filter: {
    filterBy: MessageFilters;
    filterValue: string | string[];
  }
  pagination: IPaginationOptions;
  sortBy: ISortOptions;
}

interface IMessagePayload {
    temperature: {
      value?: number;
      unit: string;
    };
    height: {
      value?: string;
      unit: string;
    };
    speed: {
      value?: string;
      unit: string;
    };
    battery: {
      value?: number;
      unit: string;
    };
    alarm: boolean;
    state: string;
    oxygen: {
      value?: number;
      unit: string;
    };
}

export interface IMessage {
  id: string;
  direction: string;
  transmittedAt: {
    unix: number;
    iso: string;
  }
  latitude: string;
  longitude: string;
  sensorId: string;
  payload: string;
  bytes: number;
  decodedPayload?: IMessagePayload;
}

export interface IGetMessageDataValues {
  messages: IMessages;
  filteredMessagesLength: number;
  initialMessagesLength: number;
}

export type IMessages = IMessage[];
