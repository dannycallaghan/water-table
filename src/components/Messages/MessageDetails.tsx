import formatcoords from 'formatcoords';
import { formatDateTime, parseJWT } from '../../lib/utils';
import { IMessage } from '../../types/types';
import GoogleMap from '../GoogleMap';
import MessageDirectionIcon from './MessageDirectionIcon';
import { Dispatch, SetStateAction, memo } from 'react';

const MessageDetails = memo(function MessageDetails(props: {
  message: IMessage;
  handleClose: Dispatch<SetStateAction<boolean>>;
}) {
  return (
    <div className="custombp:items-start fixed left-0 top-0 z-50 flex h-[100vh] w-full justify-center overflow-scroll bg-[#000000] bg-opacity-50 text-left md:items-center">
      <div className="relative mx-2 w-full bg-white p-5 md:w-4/5">
        <div
          className="absolute right-6 top-6 cursor-pointer hover:text-red-800"
          onClick={() => props.handleClose(false)}
        >
          <svg
            xmlns="http://www.w3.org/2000/svg"
            fill="none"
            viewBox="0 0 24 24"
            strokeWidth={1.5}
            stroke="currentColor"
            className="size-6"
          >
            <path
              strokeLinecap="round"
              strokeLinejoin="round"
              d="M6 18 18 6M6 6l12 12"
            />
          </svg>
        </div>
        <h2 className="mb-4 text-base font-semibold text-gray-900">Message</h2>
        <div className="mb-4">
          <div className="mb-2 flex flex-row gap-2">
            <dl className="flex flex-grow basis-1/2 flex-col">
              <dt className="text-sm text-gray-500">ID</dt>
              <dd className="text-sm text-gray-900">{props.message.id}</dd>
            </dl>
            <dl className="mb-2 flex flex-grow basis-1/2 flex-col">
              <dt className="text-sm text-gray-500">Sensor ID</dt>
              <dd className="text-sm text-gray-900">
                {props.message.sensorId}
              </dd>
            </dl>
          </div>
          <div className="mb-2 flex flex-row">
            <dl className="flex flex-grow basis-1/2 flex-col">
              <dt className="text-sm text-gray-500">Received (UTC)</dt>
              <dd className="text-sm text-gray-900">
                {formatDateTime(props.message.transmittedAt.iso as string)}
              </dd>
            </dl>
            <dl className="flex flex-grow basis-1/2 flex-col">
              <dt className="text-sm text-gray-500">Direction</dt>
              <dd className="text-sm text-gray-900">
                <MessageDirectionIcon direction={props.message.direction} />
              </dd>
            </dl>
          </div>
          <div className="flex flex-row">
            <dl className="flex flex-grow flex-col">
              <dt className="text-sm text-gray-500">Size</dt>
              <dd className="text-sm text-gray-900">{props.message.bytes} B</dd>
            </dl>
          </div>
        </div>
        <div className="flex flex-col gap-2 md:flex-row">
          <div className="basis-1/2">
            <p className="my-2 text-sm text-gray-500">
              Location:{' '}
              <span className="text-gray-900">
                {formatcoords(
                  Number(props.message.latitude),
                  Number(props.message.longitude)
                ).format()}
              </span>
            </p>
            <GoogleMap
              pins={[
                {
                  lat: props.message.latitude as string,
                  lng: props.message.longitude as string,
                  sensorId: props.message.sensorId,
                },
              ]}
              height="300px"
              width="100%"
              zoom={8}
            />
          </div>
          <div className="basis-1/2">
            <p className="my-2 text-sm text-gray-500">Payload:</p>
            <pre className="h-[300px] overflow-scroll rounded border border-gray-200 bg-gray-100 p-2 text-sm">
              {JSON.stringify(parseJWT(props.message.payload), null, 2)}
            </pre>
          </div>
        </div>
      </div>
    </div>
  );
});

export default MessageDetails;
