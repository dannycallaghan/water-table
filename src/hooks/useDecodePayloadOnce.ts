import { parseJWT } from '../lib/utils';
import { IMessages } from '../types/types';

function useDecodePayloadOnce(source: IMessages): IMessages {
  return source.map(message => ({
    ...message,
    decodedPayload: parseJWT(message.payload),
  }));
}

export default useDecodePayloadOnce;