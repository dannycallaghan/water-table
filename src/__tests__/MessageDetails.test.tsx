import sensorData from '../data/river-sensor-data-test.json';
import { expect, test, describe } from 'vitest';
import { render } from '@testing-library/react';
import MessageDetails from '../components/Messages/MessageDetails.tsx';
import useDecodePayloadOnce from '../hooks/useDecodePayloadOnce.ts';
import { IMessage } from '../types/types.ts';

describe('MessageDetails', () => {
  test('renders correctly', () => {
    const setState = () => {};
    const data = useDecodePayloadOnce(sensorData);
    const { asFragment } = render(
      <MessageDetails handleClose={setState} message={data[0] as IMessage} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
