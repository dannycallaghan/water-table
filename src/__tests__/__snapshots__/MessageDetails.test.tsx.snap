// Vitest Snapshot v1

exports[`MessageDetails > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="custombp:items-start fixed left-0 top-0 z-50 flex h-[100vh] w-full justify-center overflow-scroll bg-[#000000] bg-opacity-50 text-left md:items-center"
  >
    <div
      class="relative mx-2 w-full bg-white p-5 md:w-4/5"
    >
      <div
        class="absolute right-6 top-6 cursor-pointer hover:text-red-800"
      >
        <svg
          class="size-6"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 18 18 6M6 6l12 12"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </div>
      <h2
        class="mb-4 text-base font-semibold text-gray-900"
      >
        Message
      </h2>
      <div
        class="mb-4"
      >
        <div
          class="mb-2 flex flex-row gap-2"
        >
          <dl
            class="flex flex-grow basis-1/2 flex-col"
          >
            <dt
              class="text-sm text-gray-500"
            >
              ID
            </dt>
            <dd
              class="text-sm text-gray-900"
            >
              c19706c4-707b-4d42-9654-93248fd8d34c
            </dd>
          </dl>
          <dl
            class="mb-2 flex flex-grow basis-1/2 flex-col"
          >
            <dt
              class="text-sm text-gray-500"
            >
              Sensor ID
            </dt>
            <dd
              class="text-sm text-gray-900"
            >
              821ef78d-4b53-4bf4-83d5-686ad81deb48
            </dd>
          </dl>
        </div>
        <div
          class="mb-2 flex flex-row"
        >
          <dl
            class="flex flex-grow basis-1/2 flex-col"
          >
            <dt
              class="text-sm text-gray-500"
            >
              Received (UTC)
            </dt>
            <dd
              class="text-sm text-gray-900"
            >
              16 Jun 2019 15:00:25
            </dd>
          </dl>
          <dl
            class="flex flex-grow basis-1/2 flex-col"
          >
            <dt
              class="text-sm text-gray-500"
            >
              Direction
            </dt>
            <dd
              class="text-sm text-gray-900"
            >
              <svg
                class="size-4 text-blue-600"
                fill="none"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19.5 13.5 12 21m0 0-7.5-7.5M12 21V3"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </dd>
          </dl>
        </div>
        <div
          class="flex flex-row"
        >
          <dl
            class="flex flex-grow flex-col"
          >
            <dt
              class="text-sm text-gray-500"
            >
              Size
            </dt>
            <dd
              class="text-sm text-gray-900"
            >
              288 B
            </dd>
          </dl>
        </div>
      </div>
      <div
        class="flex flex-col gap-2 md:flex-row"
      >
        <div
          class="basis-1/2"
        >
          <p
            class="my-2 text-sm text-gray-500"
          >
            Location: 
            <span
              class="text-gray-900"
            >
              51° 12′ 19.52640″ N 1° 22′ 9.61320″ W
            </span>
          </p>
          <div
            data-testid="map"
            style="width: 100%; height: 300px; position: relative; z-index: 0;"
          />
        </div>
        <div
          class="basis-1/2"
        >
          <p
            class="my-2 text-sm text-gray-500"
          >
            Payload:
          </p>
          <pre
            class="h-[300px] overflow-scroll rounded border border-gray-200 bg-gray-100 p-2 text-sm"
          >
            {
  "temperature": {
    "value": 3,
    "unit": "C"
  },
  "height": {
    "value": "4.13",
    "unit": "M"
  },
  "speed": {
    "value": "3.26",
    "unit": "M/s"
  },
  "battery": {
    "value": 98,
    "unit": "%"
  },
  "alarm": false,
  "state": "Rising",
  "oxygen": {
    "value": 50,
    "unit": "mg/L"
  }
}
          </pre>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
